### PR TITLE
Implement `private(this)` visibility attribute

### PR DESCRIPTION
**Pull request by dkorpel**
Based on recent discussions in the newsgroup about `private` to the class instead of module:

[`restricted` member variables](https://forum.dlang.org/post/ccqkjqeueuapanfrbdxj@forum.dlang.org)
[Adding a new design constraint to D](https://forum.dlang.org/post/sldtyzrqpsqgobsgcmpy@forum.dlang.org)

The `private(this)` syntax is inspired by Scala's `private[this]` and D's `package(name)`. 
It's only enabled under the `-preview=privateThis` switch since it's not part of the language yet.

### CODE REVIEW

**Comment by ibuclaw** [comment]
> The `private(this)` syntax is inspired by Scala's `private[this]` and D's `package(name)`. It's only enabled under the `-preview=privateThis` switch since it's not part of the language yet.

Unlike Scala, this compiles when it shouldn't though:
```
class C
{
    private int secret;

    void thisTest()
    {
        this.secret = 1; // OK
    }

    void otherTest(C other)
    {
        other.secret = 2; // OK, but expect FAIL, as we are in the
                          // instance method of `this`, not `other`.
    }
}
```

**Review by ibuclaw — CHANGES_REQUESTED** [review]
As per comment, I think this should be considered as well.

**Comment by dkorpel** [comment]
> As per comment, I think this should be considered as well.

Do you think that behavior is useful, or is the point that the syntax is misleading when it doesn't do the same as Scala? I didn't see anyone ask for that behavior.

**Comment by 12345swordy** [comment]
> > As per comment, I think this should be considered as well.
> 
> Do you think that behavior is useful, or is the point that the syntax is misleading when it doesn't do the same as Scala? I didn't see anyone ask for that behavior.

I don't see that behavior is useful, as I came from the c# background not Scala. Regardless a DIP for this in order, as it not quite clear on what it should do.

**Comment by dkorpel** [comment]
> as it not quite clear on what it should do.

You just said it yourself:

> It should be accessible within the body of the class in which the private(this) is declared at.

Notice that this is a preview feature, similar to `-preview=rvaluerefparam`. It's not a definitive language addition or commitment to a specific design.

**Comment by ntrel** [comment]
C++ also allows access to private members of a (declared) parameter of type X from within a method of X.

**Comment by maxhaton** [comment]
In line with the forum thread I don't think I've ever actually seen a bug that would've been fixed by this.

File: changelog/private-this.dd [hunk 1]
```diff
@@ -0,0 +1,24 @@
+Add experimental `private(this)` visibility attribute
+
+The `private` visibility attribute only limits access to the current module.
+While `class` encapsulation [can be ensured using package modules](https://dlang.org/blog/2018/11/06/lost-in-translation-encapsulation),
+some users dislike having to put classes into individual files to accomplish this.
+
+With the `-preview=privateThis` switch, member variables can now be marked invisible outside of the
+`class`, `struct` or `union` it's declared in.
+Note that this feature it not accepted as part of the D language yet.
```
Comment (maxhaton) [inline changelog/private-this.dd:9]:
```suggestion
Note that this feature has not been accepted into the D language yet.
```

**Review by Geod24 — CHANGES_REQUESTED** [review]
Likewise, I don't see this as an useful addition to the language. The unit of encapsulation in D is the module, and the module == file, package == directory is one of the core assumption of the language. Trying to circumvent it will make the language worse, not better.

**Comment by 12345swordy** [comment]
>Trying to circumvent it will make the language worse, not better.

You need to demonstrate this instead of asserting it. The one module per file restriction causes the workaround of putting a class into its own file just to have private follow the traditional OOP paragram.
Many of us do not like that workaround. You rather get rid of the one module per file restriction?

> > as it not quite clear on what it should do.
> 
> You just said it yourself:
> 
> > It should be accessible within the body of the class in which the private(this) is declared at.
> 
> Notice that this is a preview feature, similar to `-preview=rvaluerefparam`. It's not a definitive language addition or commitment to a specific design.

Not quite, I have no idea on what is on your mind with it comes private(this) in terms of rules. A brief description on how it should work would be helpful here.

**Comment by ibuclaw** [comment]
> It should though. It should be accessible within the body of the class in which the private(this) is declared at. 

Why should it? How is accessing other instances privates a good idea?

> Otherwise how are you supposed to create a deep copy constructor?
> 

Then private is not the visibility you are looking for.

**Comment by mdparker** [comment]
> ou need to demonstrate this instead of asserting it. 

@12345swordy That's backwards. Whether his assertion is right or wrong, it's up to those pushing this feature to demonstrate its benefit.

> the workaround of putting a class into its own file just to have private follow the traditional OOP paragram

That's not a workaround. That's the design of D. As has been pointed out in the forums, if something shouldn't have access to an aggregate's innards, then they don't belong in the same module. The concept is the same for class internals in Java, but Java doesn't have D-style modules. We've just moved it up one level. The module is the unit of encapsulation, not the class. The module is part of the private interface of any aggregate defined in the module. That's the way to look at this.

However, there is a hole. Because the module is part of the private interface, then the module's public interface is effectively part of the aggregate's public interface. And as has been pointed out in the forums, that allows one to bypass invariant checks and mutex locks (on synchronized classes). But a `private(this)` doesn't fully plug that hole, as any `private` member will still be vulnerable to it.

**Comment by 12345swordy** [comment]
> > It should though. It should be accessible within the body of the class in which the private(this) is declared at.
> 
> Why should it? How is accessing other instances privates a good idea?

Like I said before: Creating a deep copy constructor. It does not make sense that a class can't access other instance private when they share the same type definition here. I.E. Class A should access the private instance of an class A variable.   I am not talking about two different types here.
 
> > Otherwise how are you supposed to create a deep copy constructor?
> 
> Then private is not the visibility you are looking for.
Nonsense, I am operating from a C# background version of private not a Scala background version of private.

**Comment by mdparker** [comment]
> Nonsense, I am operating from a C# background version of private not a Scala background version of private.

And we're operating from a D version of private. Looks like there's no standard definition after all.

**Comment by 12345swordy** [comment]
> > ou need to demonstrate this instead of asserting it.
> 
> @12345swordy That's backwards. Whether his assertion is right or wrong, it's up to those pushing this feature to demonstrate its benefit.
> 
> > the workaround of putting a class into its own file just to have private follow the traditional OOP paragram
> 
> That's not a workaround. 
YES, IT IS. It been brought up in the forums already here. No need for repeated explanations here.

**Comment by 12345swordy** [comment]
@mdparker 
"if something shouldn't have access to an aggregate's innards, then they don't belong in the same module."
non sequitur. The conclusion does not follow the premises. Replace module with File/Dictionary/Computer to see the absurdity.

**Comment by ibuclaw** [comment]
> > As per comment, I think this should be considered as well.
> 
> Do you think that behavior is useful, or is the point that the syntax is misleading when it doesn't do the same as Scala? I didn't see anyone ask for that behavior.

It is logically consistent with package(Package) at least, if you consider the missing link between the two being private(Class).

**Comment by 12345swordy** [comment]
> @12345swordy No, it isn't a non sequitur.

Yes it is. The statement by itself is a non sequitur.

> In D, the module is part of the aggregate's private interface. 

Which can be change via a DIP.

**Comment by ibuclaw** [comment]
> @mdparker 
> "if something shouldn't have access to an aggregate's innards, then they don't belong in the same module."
> non sequitur. The conclusion does not follow the premises. Replace module with File/Dictionary/Computer to see the absurdity of the statement. 

"If something shouldn't have access to a private key, then they don't belong in the same file/dictionary/computer."

Makes perfect sense to me.

**Comment by 12345swordy** [comment]
I can keep seeing the false dichotomy when it comes to files usage here. It not an all or nothing when it comes to file usage here, there is a thing called partial file access by corporation with other people here. If you can't understand that notation here, then I am done with D here.

**Comment by maxsamukha** [comment]
> In line with the forum thread I don't think I've ever actually seen a bug that would've been fixed by this.

I can't remember I've ever seen a bug that would've been fixed by module-private. By that argument, why have 'private' at all if we could just leave private members undocumented or document them as 'don't touch this'? Python goes with '_', and nobody complains.

No matter how I dislike module-private, I would leave it alone for these reasons:

1) The debate about whether `private(this)` should be class-level or object-level will go forever, and nobody will be happy anyway; and having both `private(this)` and `private(class)` is a no go.

2) Object-level would not be specified and implemented reasonably:

```d
class C
{
    void otherTest(C other)
    {
        other = this;
        other.secret = 2; // should pass?
    }
}
```

3) Class-level would have the same issues with 'invariant/synchronized' as module-level does:
```
synchronized class C
{
    // lock 'others'? in which order?
    void otherTest(C[] others)
    {
    }
}
```

**Comment by ibuclaw** [comment]
> > In line with the forum thread I don't think I've ever actually seen a bug that would've been fixed by this.
> 
> I can't remember I've ever seen a bug that would've been fixed by module-private. By that argument, why have 'private' at all if we could just leave private members undocumented or document them as 'don't touch this'? Python goes with '_', and nobody complains.
> 
> No matter how I dislike module-private, I would leave it alone for these reasons:
> 
> 1) The debate about whether `private(this)` should be class-level or object-level will go forever, and nobody will be happy anyway; and having both `private(this)` and `private(class)` is a no go.
> 
> 2) Object-level would not be specified and implemented reasonably:
> 

If there is no implicit or explicit `this` dereference, then reject it. Var lifetime/borrow tracking doesn't have to be involved in the process - especially given that `@live` is still a moving target.

> 
> 3) Class-level would have the same issues with 'invariant/synchronized' as module-level does:

Right, so object-level would mean that there'll be no need to generate invariants for the object-private fields, as they would be invisible even in reflection code (getMembers). It would be wishful thinking to say the same for synchronized, but for shared references to the same instance.

**Comment by dkorpel** [comment]
> I don't see this as an useful addition to the language

This PR doesn't add this to the language, it adds an experimental preview feature. Please review it as that, and keep discussion of the merits of the idea on the forums so the comments here can focus on reviewing the implementation.

**Pull request closed without merging**